File: trabalho4_links/ruby_api/linkextractor.rb
require 'sinatra'
require 'open-uri'
require 'nokogiri'
require 'json'
require 'redis'

set :bind, '0.0.0.0'
set :port, 5000

redis_url = ENV['REDIS_URL']
cache = (redis_url && !redis_url.empty?) ? Redis.new(url: redis_url) : nil

get '/api/links' do
  url = params['url']
  if cache
    begin
      cached = cache.get(url)
      return cached if cached
    rescue
    end
  end
  begin
    doc = Nokogiri::HTML(URI.open(url, "User-Agent" => "Mozilla/5.0", :read_timeout => 10))
    links = doc.css('a').map { |a| a['href'] }.compact
    result = { url: url, links: links }.to_json
    if cache
      begin cache.setex(url, 300, result) rescue nil end
    end
    content_type :json
    result
  rescue => e
    status 400
    { error: e.message }.to_json
  end
end

File: trabalho4_links/python_api/linkextractor.py
import os, redis, requests
from bs4 import BeautifulSoup
from flask import Flask, request, jsonify

app = Flask(__name__)
redis_url = os.environ.get('REDIS_URL')
cache = redis.from_url(redis_url) if redis_url else None

@app.route('/api/links')
def links():
    url = request.args.get('url')
    if cache:
        try:
            cached = cache.get(url)
            if cached: return cached
        except: pass
    try:
        res = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'}, timeout=10)
        soup = BeautifulSoup(res.text, "html.parser")
        links = [a.get('href') for a in soup.find_all('a') if a.get('href')]
        result = jsonify({"url": url, "links": links})
        if cache:
            try: cache.set(url, result.get_data(), ex=300)
            except: pass
        return result
    except Exception as e:
        return jsonify({"error": str(e)}), 400

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000)
